Handle oidc in spring cloud client

glefloch opened the pull request:
This is a draft PR that introduce a new extension on top of the existing `quarkus-spring-cloud-config-client` that allows the connection to a `spring-cloud-config-server` the OIDC enabled. 

This is one of my first extension, all comments appreciated :)

close #17642

glefloch committed:
Handle oidc in spring cloud client

geoand commented on extensions/spring-cloud-config-client-oidc/deployment/src/main/java/io/quarkus/spring/cloud/config/client/SpringCloudConfigClientOidcProcessor.java:
```diff
@@ -0,0 +1,23 @@
+    @BuildStep
+    @Record(ExecutionTime.RUNTIME_INIT)
+    public void configure(BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
+        additionalBeans.produce(AdditionalBeanBuildItem.unremovableOf(SpringCloudConfigClientOidcProcessor.class));
```
I think you mean `SpringCloudClientBasicAuthProvider` here, right?

geoand replied:
Actually I think you've added it later so I think this method is entirely redundant

glefloch replied:
I should rename the package to add oidc but this is the step that generated the OIDc provider. In the other module, I generate the basicAuth provider.

geoand reviewed (commented):
Thanks!

The Spring stuff looks good to me.

Could we also have a test for this?

gsmet commented:
Might be a stupid question but do we really need a separate extension? I was wondering if using optional dependencies and config could be enough?

geoand commented:
Yeah, I had the same question myself. But in that case we would need to make this opt-in because a user could easily be using oidc for other parts and not want it for the Spring Cloud Config Server auth part

glefloch commented:
@geoand I added a test but I'm struggling with an error. As soon as I try to use the `OidcClients` object  to get the correct `OidcClient` (https://github.com/quarkusio/quarkus/pull/18070/files#diff-127f1247c508778501d11cad6eda21970f5bf9256bd1dfcf2e6be970fb0eb707R24)
I get the following error as the `OidcClient` does not seems to be initialized yet:  
```
2021-07-07 11:50:54,573 ERROR [io.qua.spr.clo.con.cli.run.SpringCloudConfigServerClientConfigSourceProvider] (main) Unable to obtain configuration from Spring Cloud Config Server at http://localhost:8089/base: javax.enterprise.inject.CreationException: Synthetic bean instance for io.quarkus.oidc.client.OidcClients not initialized yet: io_quarkus_oidc_client_OidcClients_97d170e1550eee4afc0af065b78cda302a97674c
        - a synthetic bean initialized during RUNTIME_INIT must not be accessed during STATIC_INIT
        - RUNTIME_INIT build steps that require access to synthetic beans initialized during RUNTIME_INIT should consume the SyntheticBeansRuntimeInitBuildItem
        at io.quarkus.oidc.client.OidcClients_777be49f5563055569a690ceca758ae0746df06b_Synthetic_Bean.create(OidcClients_777be49f5563055569a690ceca758ae0746df06b_Synthetic_Bean.zig:161)
        at io.quarkus.oidc.client.OidcClients_777be49f5563055569a690ceca758ae0746df06b_Synthetic_Bean.create(OidcClients_777be49f5563055569a690ceca758ae0746df06b_Synthetic_Bean.zig:184)
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)

```

I try adding a `@Consume(SyntheticBeansRuntimeInitBuildItem.class)` but then, I get circular dependency exception. Have you an idea of how I could solve this?

geoand commented:
Hm, actually we won't be able to solve it. 
The bootrstap config stuff MUST happen before any runtime init code, which means you can't access such synthetic beans that are runtime init.
I assume that the OidcClients synthetic bean is runtime init for a reason, although I don't know anything about it to be honest :)
We'll have to find some other way to interact with OIDC

glefloch commented:
Thanks for your answer @geoand, maybe @sberyozkin knows more about `OidcClient` being runtime init. 

I will start to implement a simple bean that gets the `access_token` directly in the extension.

geoand commented:
FWIW, this is a typical kind of chicken and egg problem that unfortunately doesn't always have a great solution :(

glefloch commented:
Right, configuration needs authentication but authentication needs configuration :)

sberyozkin commented:
Hi @glefloch @geoand `OidcClient `is runtime initiialized because its connection properties can be overridden at runtime with the system properties, etc.
@glefloch may be you can use `Arc` to get the required `OidcClient` during the request time ?

If you prefer to go some other way then it is totally fine with me, thanks

glefloch commented:
@geoand @sberyozkin I simplified things and removed the new extension. I wasn't able to get something working due to init constraints. 
I moved everything in the `spring-cloud-config-client` extension. I manually create a `OidcClientImpl` class based on configuration. 
However, I kept the `integration-tests/spring-cloud-config-client-oidc` module as it requires a keycloak resource.

geoand commented on devtools/bom-descriptor-json/pom.xml:
```diff
@@ -2476,6 +2476,19 @@
                         </exclusion>
                     </exclusions>
                 </dependency>
+                <dependency>
```
This needs to be removed, right?

geoand commented on devtools/bom-descriptor-json/pom.xml:
```diff
@@ -2476,6 +2476,19 @@
                     </exclusions>
                 </dependency>
+                <dependency>
+                    <groupId>io.quarkus</groupId>
```
Same here

geoand commented on docs/pom.xml:
```diff
@@ -2437,6 +2437,19 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
```
And here

geoand commented on extensions/spring-cloud-config-client/deployment/pom.xml:
```diff
@@ -29,6 +29,10 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-vertx-core-deployment</artifactId>
         </dependency>
+        <dependency>
```
And here

glefloch commented:
Oh right @geoand, I pushed an update.

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/Feature.java:
```diff
@@ -141,6 +141,7 @@
     SPRING_BOOT_PROPERTIES,
     SPRING_CACHE,
     SPRING_CLOUD_CONFIG_CLIENT,
+    SPRING_CLOUD_CONFIG_CLIENT_OIDC,
```
We probably don't need this now either, right?

glefloch replied:
Right, sorry for that..

geoand commented on extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientConfig.java:
```diff
@@ -113,6 +114,12 @@
     public Map<String, String> headers;
 
+    /**
+     * Name of the OIDC client to use. Use default client by default.
```
```suggestion
     * Name of the OIDC client to use. Uses the default client if not set.
```

geoand commented:
Thanks a lot @glefloch! 

The integration looks good, but I'll leave it to @sberyozkin to approve as I can't really comment on the OIDC stuff.

geoand commented:
The CI failure seems related

glefloch commented:
Test looks ok but I have the following error in the `quarkus-test-extension-nativetests` module : 

```
Caused by: java.lang.ClassNotFoundException: org/apache/xalan/processor/TransformerFactoryImpl
    at java.lang.Class.forName0 (Native Method)
    at java.lang.Class.forName (Class.java:398)
    at javax.xml.transform.FactoryFinder.getProviderClass (FactoryFinder.java:122)
    at javax.xml.transform.FactoryFinder.newInstance (FactoryFinder.java:165)
    at javax.xml.transform.FactoryFinder.find (FactoryFinder.java:212)
    at javax.xml.transform.TransformerFactory.newInstance (TransformerFactory.java:126)
    at org.atteo.xmlcombiner.XmlCombiner.buildDocument (XmlCombiner.java:236)
    at io.quarkus.extest.deployment.UberJarConfigBuildStep.uberJarMergedResourceBuildItem (UberJarConfigBuildStep.java:47)
```

@geoand do you know where it comes from?

geoand commented:
I have no idea unfortunately

sberyozkin commented on extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/credentials/SpringCloudConfigClientOidcCredentialsProvider.java:
```diff
@@ -0,0 +1,110 @@
+                    .onItem().transform(json -> json.getString("token_endpoint")).await().indefinitely();
+        }
+
+        String grantType = oidcClientConfig.grant.getType().getGrantType();
```
Hi @glefloch  Spring Cloud must be supporting a specific grant type only, probably `client_credentials` - `OidcClient` needs to support different grants - but it will fail here unless it is the grant supported by Spring Cloud, so IMHO it makes sense to throw a configuration error if it is not the one recognized by Spring Cloud

sberyozkin commented on extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/credentials/SpringCloudConfigClientOidcCredentialsProvider.java:
```diff
@@ -0,0 +1,110 @@
+        return grantParams;
+    }
+
+    private static MultiMap getTokenGrantParam(OidcClientConfig oidcConfig, String grantType) {
```
This is probably a redundant code for Spring Cloud

sberyozkin commented:
Hi @glefloch sorry for a delay, I was on PTO and also missed the latest updates. 
IMHO it looks good, but I'm pretty sure most of that code to do with initializing `OidcClientImpl` can be removed given that using a token with `SpringCloud` represents a very specific case (specifc grant, specific authentication - client id and secret are either passed as `Authorization Bearer` or as form parameters) etc, so please check what SpringCloud expects and remove most of that code :-)
I'd also consider having a significantly simpler OidcClient configuration support in scope of this extension as most of that configuration is meant to cover various cases. Perhaps you can simply have 
```
SpringCloudOidcClient extends OidcCommonConfig {
}
```

and may be add `refreshTokenTimeSkew` if will be needed over time.

sberyozkin commented on extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/credentials/SpringCloudConfigClientOidcCredentialsProvider.java:
```diff
@@ -0,0 +1,110 @@
+
+    @Override
+    public void addAuthenticationInfo(HttpRequest<?> request) {
+        request.bearerTokenAuthentication(oidcClient.getTokens().await().indefinitely().getAccessToken());
```
I'm not 100% sure requesting a new token on every authentication request will work with the Spring Cloud provider - as there are usually some limits on how many tokens a given client can acquire.

The produced OidcClient in `quarkus-oidc-client` will cache the token and handle auto-refreshing it once they have expired - but if the client is used directly then the token has to be refreshed if it has expired.

I think you should use `io.quarkus.oidc.client.runtime.TokensHelper` instead (this is used by OidcClient producers) - pass it the initialized OidcClient - and it will take care of the rest (caching, refreshing, etc)

glefloch replied:
You are right, I used the `TokensHelper` here. However, I think this is called only once when the application starts, it load config then, it will only be refreshed on the next restart.

sberyozkin replied:
@glefloch Hi, I meant the access token will be refreshed if needed - it will be acquired and then reused while it is active and then refreshed when needed...

glefloch commented:
Thanks @sberyozkin for the review. Looking at Spring Cloud config, there is not that much on OIDC support. Actually, to start a config server, you need to create a spring boot app, with the `spring-cloud-config-server` starter, and for security you can do what you want (based on `spring-security`). But I agree with you that in most case, it will be a `client_credentials` grant type. For a first version, I can limit to `client_credentials` grant type, WDYT?

sberyozkin commented:
@glefloch 
> For a first version, I can limit to client_credentials grant type, WDYT?

I've looked here,
https://github.com/ka4ok85/spring-cloud-config-client-jwt

may be the `password` grant which is used... I'm not 100% sure, may be it is simpler to keep the similar config as OidcClient....

gsmet commented:
This one now conflicts. I'm not sure if we should pursue or close.

What's your take @geoand @sberyozkin @glefloch ?

geoand commented:
I think we can close it as we have not had any real demand for it

glefloch commented:
Yes I agree too.

geoand commented:
Thanks for your effort in any case @glefloch!
